### ISSUE TITLE
Track Number in Song Title

### ISSUE THREAD

**Issue by CTParkinson** [airsonic/airsonic#1608]
## Problem description

Seems that if the song title starts with the track number then the track number is stripped from the song title.
Which for most cases seems fine, but I noticed that for "10 Years Today", which is track 10 on [The Poison](https://en.wikipedia.org/wiki/The_Poison#Track_listing), that the 10 is removed and I can't make it not truncate the title:
![image](https://user-images.githubusercontent.com/46909431/78673044-7c7ca000-78d9-11ea-922c-55e9706b7e24.png)

Have checked all data in the file, it is all correct.
The filename itself is "10 - 10 Years Today.mp3" so I'm not sure if it's a greedy regex or something like that.
Displays like that in the album listing, and also the Now Playing on the right side, and in all results through the API, so I assume it's done as it is entered into the database?
But I don't know where to begin looking into the source to potentially offer a fix :(

Presumably it is the same for any other track that has its title coincidentally begins with the track number? Can't recall any other songs where this may be the case to check those unfortunately.

### Steps to reproduce

1. Have a song title correctly begin with its track number.
2. View that song in any view.
3. Observe incorrectly truncated song title.

## System information

 * **Airsonic version**: 10.5.0-RELEASE – November 7, 2019
 * **Client**: Airsonic's own UI, and custom interface via API.

**Comment by randomnicode** [comment]
Hello
> But I don't know where to begin looking into the source to potentially offer a fix :(

The MediaScannerService and the MediaFileService along with the MediaFileDao will be your best bet. They enter the data into the DB. I think you're particularly interested in the createMediaFile() method of the MediaFileService. The MetaDataParsers that are used to pull in the data for the title might be interesting too. Overtly I dont see any issues but I haven't looked too closely. You might want to check your db first to see if the db has the wrong title (table: `media_file`, column: `title`)

**Comment by CTParkinson** [comment]
I'll happily take a look, but have also never written a line of Java, so unless it's a simple fix it's probably unlikely that I'll have an elegant solution.

Also, if I edit tags and set the correct title;
![image](https://user-images.githubusercontent.com/46909431/78704375-ac409d80-7903-11ea-86d6-96bf3f8eeb10.png)
it marks as updated, but the issue persists.

**Comment by randomnicode** [comment]
> So I guess it's run twice at somepoint?

Hehe. No. It's actually simpler than that. The title coming in isn't the filename, it's whatever is parsed from the id3 tags. So your title coming in is `10 Years Today`, not `10 - 10 Years Today`. When you apply your regex to that (the former), you see how it lops off the `10 ` to make `Years Today`.

This seems a particularly uncomely edge case, because in 99.99% of the times I _think_ we do want to lop it off, but I'm not sure, so I'm hoping to think aloud to reach a consensus and a conclusive fix.

The logic here says:
- If your track number is null: don't lop it off the title (the track is null so there's nothing to lop off)
- If title doesn't start with the track number: don't lop it off (again, you have nothing to lop off)
- Otherwise the track number and the begininning of the title match: lop it off because it's likely redundant

This is reasonable in most cases because most albums kind of exist like this:
| id3 trackNum | filename |
|---|---|
| 10 | 10 Song.mp3 |

so lopping it off makes sense.

However, if your title is actually coming from the id3 tags themselves instead of the filename, it makes me wonder if we can use that as the key distinction we need to introduce. But this distinction will also fail if the following scenario is common:

| id3 trackNum | filename | id3 title |
|--------|-------|----|
| 10   | 10 Song A.mp3 | 10 Song A

So in short, what I propose is this:
- Figure out if the title is coming from the filename or the id3 tag
- if it's coming from the filename, it's open hunting season, lop off until kingdom come. It probably deserved it.
- If it's coming from the id3 tag, it's sacred. Don't even think about touching it. Kiss the metaphorical ground it walks on. It gets a free expedited pass to the database, no questions asked.

However this is only feasible if there are no scenarios like Table B above, because the fix treats id3 tags as immutable. If this fix goes in and the situation above is common, then there will be an armada of angry rioters wondering why a virus wasn't enough and now they have to additionally contend with a number in their titles.

So in short, what I'm asking is: how common is the above scenario? Do ID3 tags ever have track numbers in their title fields? People who do more ripping would presumably know what the ripping standards are nowadays.

**Comment by CTParkinson** [comment]
I can't imagine it's at all common. And I'm okay with it removing the '10' from the ID3 tag title, that makes sense.
So let's move the goal post. It would be nice when I am editing tags via the Airsonic Edit Tags interface if it trusted my input.

The ID3 tag doesn't need to be sacred but, if I tell Airsonic what it should be, then ignoring my input and doing what it wants is a little frustrating.
I can handle having to modify one track each time it scans my media (if I remember).

But I can't see where it does that however.
`editTags.jsp` calls `setTags` in TagService, TagService calls `getMetaData` on the parser that truncates the metadata title, but then the title is overwritten by `setTitle` on the `newMetaData` object and that looks like it takes the title input as it is written and presumably that's it?
But at no point does it actually set the title that I have written.

The only thing I can see, and I might be way off base, is that the body of `setMetaData` in the DefaultMetaDataParser (which I assume is the one being used) is empty.
https://github.com/airsonic/airsonic/blob/fa1a700ebdeede00b919a1685818d9ec6139b2b6/airsonic-main/src/main/java/org/airsonic/player/service/metadata/DefaultMetaDataParser.java#L69-L70
And so it's not being updated at all..?

But I don't know enough about the project to determine which parser is actually being used, though that aligns with my experience in this issue. Even though I have updated via the Edit Tags page before and now I'm confusing myself...

**Comment by MightyPork** [comment]
I often find songs with the track number embedded in their ID3 tags when downloading music, so there is a legitimate argument for keeping it. Perhaps even more common is song titles starting with  "- " because the tagging software joined empty string with track title by hyphen, for some reason.

You could also just rename the song to use "Ten" instead of "10" or perhaps add apostrophe or something else before the number

**Comment by CTParkinson** [comment]
So as a temporary solution I just set the tags title to "10 - 10 Years Today" through the Edit tag page, and let it strip the first one as it does and now it displays correctly, until the next time.

Having it manipulate my input is not desirable, but the frequency at which this would occur makes it almost negligible...

**Comment by randomnicode** [comment]
> The only thing I can see, and I might be way off base, is that the body of setMetaData in the DefaultMetaDataParser (which I assume is the one being used) is empty.

It's actually not being used. It's used as a last resort of sorts when the world has failed and it needs to save the day (by doing nothing). What's actually being used is the JaudiotaggerParser (another implementation of the MetaDataParser), where you may find this gem:

https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/service/metadata/JaudiotaggerParser.java#L230

So, in short you are editing, but the question is _what_ are you editing. You presume that you're editing the database when, in fact, you're editing the ID3 tag.
So let's say you type "10 Years Today". ID3 tag gets updated to 10 Years Ago. The system then does its scan which strips out the 10 again from the ID3 tag and store it in the DB as Years Today, which then it presents itself to you in the table.

So the issue here seems to be: we update the ID3 tags, but read from the DB when populating the edit tags table. Maybe there should be a switch or something to denote what you're updating.

Unfortunately in your case, now your ID3 tag probably reads an unwanted "10 - 10 Years Today"

**Comment by randomnicode** [comment]
> I often find songs with the track number embedded in their ID3 tags when downloading music, so there is a legitimate argument for keeping it

Fair enough, so my solution doesn't work in this case. But we should probably at least have an option to select whether the db or the id3 tag is being updated
